go/merkletree_new/merkle_tree_interface.go

```diff
@@ -0,0 +1,65 @@
+package merkletree
+
+// MerkleTreeInterface is the methods which are required to manipulate the
+// contents of a Merkle Hash Tree (that is, add new entries), and to query
+// the tree for inclusion and transparency proofs.
+type MerkleTreeInterface interface {
+	// CurrentRootHash returns the hash of the entire current tree.
+	CurrentRootHash func () ([]byte, error)
+
+	// InclusionProof returns a list of the hashes of the "sibling"
+	// nodes to the leaf entry located at the (0-based) index specified.
+	// The returned list is presented in order, starting from the
+	// sibling to the item itself, and ending with a hash that is an
+	// immediate child of the root.
+	InclusionProof func (entry uint64) ([]byte, error)
+
+	// ConsistencyProof returns a list of the hashes that make up the
+	// consistency proof between the hash trees with size |tree1| and
+	// |tree2|.
+	ConsistencyProof func (sz uint64) ([]byte, error)
+}
+
+// MerkleTreeDataInterface defines the means by which a merkle hash tree can
+// retrieve data from an arbitrary data store, without tying the tree
+// implementation to any particular backend storage technology.
+type MerkleTreeDataInterface interface {
```

Comment:
These two interfaces are an interesting idea, I'd only ever imagined the tree being all in-memory.
Will be interesting to see what the trade-off will be in terms of space efficiency, but this could be very nice for insane tree sizes.